// @heroui/react/dist/components/calendar/calendar.js
"use client";
import { calendarVariants } from '@heroui/styles';
import React__default, { createContext, useContext } from 'react';
import { Calendar, CalendarCell as CalendarCell$1, CalendarHeaderCell as CalendarHeaderCell$1, CalendarGridHeader as CalendarGridHeader$1, CalendarGrid as CalendarGrid$1, CalendarGridBody, Button, Heading } from 'react-aria-components';
import { composeTwRenderProps, composeSlotClassName } from '../../utils/compose.js';
import { jsx, jsxs } from 'react/jsx-runtime';

const CalendarContext = /*#__PURE__*/createContext({});

/* -------------------------------------------------------------------------------------------------
 * Calendar Root
 * -----------------------------------------------------------------------------------------------*/

function CalendarRoot({
  children,
  className,
  isDisabled,
  ...rest
}) {
  const slots = React__default.useMemo(() => calendarVariants({
    isDisabled
  }), [isDisabled]);
  return /*#__PURE__*/jsx(CalendarContext, {
    value: {
      slots
    },
    children: /*#__PURE__*/jsx(Calendar, {
      "data-slot": "calendar",
      isDisabled: isDisabled,
      ...rest,
      className: composeTwRenderProps(className, slots.base()),
      children: children
    })
  });
}

/* -------------------------------------------------------------------------------------------------
 * Calendar Header
 * -----------------------------------------------------------------------------------------------*/

const CalendarHeader = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(CalendarContext);
  return /*#__PURE__*/jsx("header", {
    className: composeSlotClassName(slots?.header, className),
    "data-slot": "calendar-header",
    ...props,
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Calendar Heading
 * -----------------------------------------------------------------------------------------------*/

const CalendarHeading = ({
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(CalendarContext);
  return /*#__PURE__*/jsx(Heading, {
    "data-slot": "calendar-heading",
    ...props,
    className: composeSlotClassName(slots?.heading, className)
  });
};

/* -------------------------------------------------------------------------------------------------
 * Calendar Nav Button
 * -----------------------------------------------------------------------------------------------*/

const CalendarNavButton = ({
  children,
  className,
  slot,
  ...props
}) => {
  const {
    slots
  } = useContext(CalendarContext);
  return /*#__PURE__*/jsx(Button, {
    "data-slot": "calendar-nav-button",
    slot: slot,
    ...props,
    className: composeTwRenderProps(className, slots?.navButton()),
    children: children || (slot === "previous" ? "‹" : "›")
  });
};

/* -------------------------------------------------------------------------------------------------
 * Calendar Grid
 * -----------------------------------------------------------------------------------------------*/

const CalendarGrid = ({
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(CalendarContext);
  return /*#__PURE__*/jsxs(CalendarGrid$1, {
    "data-slot": "calendar-grid",
    ...props,
    className: composeSlotClassName(slots?.grid, className),
    children: [/*#__PURE__*/jsx(CalendarGridHeader, {
      children: day => /*#__PURE__*/jsx(CalendarHeaderCell, {
        children: day
      })
    }), /*#__PURE__*/jsx(CalendarGridBody, {
      children: date => /*#__PURE__*/jsx(CalendarCell, {
        date: date,
        children: date.day
      })
    })]
  });
};

/* -------------------------------------------------------------------------------------------------
 * Calendar Grid Header
 * -----------------------------------------------------------------------------------------------*/

const CalendarGridHeader = ({
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(CalendarContext);
  return /*#__PURE__*/jsx(CalendarGridHeader$1, {
    "data-slot": "calendar-grid-header",
    ...props,
    className: composeSlotClassName(slots?.gridHeader, className)
  });
};

/* -------------------------------------------------------------------------------------------------
 * Calendar Header Cell
 * -----------------------------------------------------------------------------------------------*/

const CalendarHeaderCell = ({
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(CalendarContext);
  return /*#__PURE__*/jsx(CalendarHeaderCell$1, {
    "data-slot": "calendar-header-cell",
    ...props,
    className: composeSlotClassName(slots?.headerCell, className)
  });
};

/* -------------------------------------------------------------------------------------------------
 * Calendar Cell
 * -----------------------------------------------------------------------------------------------*/

const CalendarCell = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(CalendarContext);
  return /*#__PURE__*/jsx(CalendarCell$1, {
    "data-slot": "calendar-cell",
    ...props,
    className: composeTwRenderProps(className, slots?.cell()),
    children: values => {
      const {
        formattedDate,
        isDisabled,
        isHovered,
        isOutsideMonth,
        isSelected,
        isUnavailable
      } = values;
      return /*#__PURE__*/jsx("div", {
        className: slots?.cellButton({
          isDisabled,
          isHovered,
          isOutsideMonth,
          isSelected,
          isUnavailable
        }),
        children: typeof children === "function" ? children(values) : children || formattedDate
      });
    }
  });
};

export { CalendarCell, CalendarGrid, CalendarGridHeader, CalendarHeader, CalendarHeaderCell, CalendarHeading, CalendarNavButton, CalendarRoot };
